README.md
# pkf-splunk

v0.4.0 Set launch parameters
v0.5.0 -> v0.6.1 had changes rolled back by v0.7.0
v0.7.0 Configure Splunk as systmed service


CHANGELOG.md
v0.5.0
  -Change from using systemd to Splunk's boot-start command coupled with splunk-launch parameters
v0.5.1
  -Remove systemd resource left in code by mistake
v0.6.0
  -Leverage "limits" cookbook instead of custom ulimit configurations
v0.7.0
  -Configure Splunk back as systmed service, removing v5 and v6 updates
v0.7.1
  -Fix missing systemd lines after merge conflict
v0.7.2
  -Removed empty "Description" option in systemd declaration
v0.8.0
  -Use "limits" recipe instead of setting them manually
v0.8.1
  -Update tests for systemd instead of init.d


test/recipes/host_setup.rb
# # encoding: utf-8

# Inspec test for recipe pkf-splunk::host_setup

# The Inspec reference, with examples and extensive documentation, can be
# found at https://docs.chef.io/inspec_reference.html

#Validate Splunk service account
describe user ('splunk001') do
  it { should exist }
  its('home') { should eq '/agents/splunkforwarder' }
  its('group') { should eq 'splgroup' }
end

#Validate THP is disabled
describe file '/sys/kernel/mm/transparent_hugepage/enabled' do
  its('content') { should match '\[never\]' }
end
describe file '/sys/kernel/mm/transparent_hugepage/defrag' do
  its('content') { should match '\[never\]' }
end

#Validate ulimits for Splunk service account
describe command ('su - splunk001 -c "ulimit -Ha"') do
  its ('stdout') { should include '(-i) 16000' }
  its ('stdout') { should include '(-u) 16000' }
  its ('stdout') { should include '(-n) 64000' }
end


test/recipes/agent_setup.rb
# # encoding: utf-8

# Inspec test for recipe pkf-splunk::agent_setup

# The Inspec reference, with examples and extensive documentation, can be
# found at https://docs.chef.io/inspec_reference.html

#Check if Splunk is running correctly
describe processes 'splunkd' do
  its('states') { should eq ['Sl'] }
  its('users') { should eq ['splunk0+'] }
end

#Check if splunk-launch is configured
describe file '/agents/splunkforwarder/etc/splunk-launch.conf' do
  its('content') { should match 'SPLUNK_OS_USER = ' }
  its('content') { should match 'SPLUNK_IGNORE_SELINUX = ' }
  its('content') { should match 'SPLUNK_FIPS = ' }
end

#Ensure setupfiles are removed
describe file '/agents/splunkforwarder_6-5-2.tgz' do
  it { should_not exist }
end

#Verify Splunk is enabled
describe file '/etc/init.d/splunk' do
  it { should exist }
end


test/recipes/default_test.rb
# # encoding: utf-8

# Inspec test for recipe pkf-splunk::default

# The Inspec reference, with examples and extensive documentation, can be
# found at https://docs.chef.io/inspec_reference.html


recipes/agent_setup.rb
#
# Cookbook Name:: pkf-splunk
# Recipe:: agent_setup
#
# Copyright (c) 2017 The Authors, All Rights Reserved.

#Downlaod Splunk installable
remote_file "#{node['splunk']['directory']}/#{node['splunk']['installable']}" do
  source "#{node['splunk']['repository']}/#{node['splunk']['installable']}"
  mode '0755'
  action :create
end

#Untar Splunk installable
execute 'Untar' do
  command "tar xzf #{node['splunk']['installable']}"
  cwd node['splunk']['directory']
  user "#{node['splunk']['user']}"
  not_if { File.exists?("#{node['splunk']['directory']}/splunkforwarder/bin/splunk") }
end

#Remove Splunk installable
remote_file "#{node['splunk']['directory']}/#{node['splunk']['installable']}" do
  action :delete
end

#Configure Splunk Deployment Server
template "#{node['splunk']['directory']}/splunkforwarder/etc/system/local/deploymentclient.conf" do
  source 'deploymentclient.conf.erb'
  owner "#{node['splunk']['user']}"
  group "#{node['splunk']['group']}"
  mode '0750'
end

#Configure Splunk to launch with FIPS, service account, and ignore SELinux
template "#{node['splunk']['directory']}/splunkforwarder/etc/splunk-launch.conf" do
  source 'splunk-launch.conf.erb'
  owner "#{node['splunk']['user']}"
  group "#{node['splunk']['group']}"
  mode '0750'
end

#Configure Splunk to start on boot w/ systemd
systemd_unit 'splunk.service' do
  content <<-EOU.gsub(/^\s+/, '')
  [Unit]
  Description=Splunk Universal Forwarder service
  After=network.target
  Wants=network.target

  [Service]
  Type=forking
  RemainAfterExit=no
  Restart=always
  RestartSec=30s
  User=#{node['splunk']['user']}
  Group=#{node['splunk']['group']}
  ExecStart=#{node['splunk']['directory']}/splunkforwarder/bin/splunk start --accept-license --answer-yes --no-prompt
  ExecStop=#{node['splunk']['directory']}/splunkforwarder/bin/splunk stop
  ExecReload=#{node['splunk']['directory']}/splunkforwarder/bin/splunk restart

  [Install]
  WantedBy=multi-user.target
  EOU
  action [ :create, :enable, :start ]
end


recipes/host_setup.rb
#
# Cookbook Name:: pkf-splunk
# Recipe:: host_setup
#
# Copyright (c) 2017 The Authors, All Rights Reserved.

#Set directory for Splunk service account
directory "#{node['splunk']['directory']}/splunkforwarder" do
  owner "#{node['splunk']['user']}"
  group "#{node['splunk']['group']}"
  mode '0755'
  recursive true
  action :nothing
end

#Create group for Splunk service account
group 'splgroup'

#Create Splunk service account
user "#{node['splunk']['user']}" do
  group "#{node['splunk']['group']}"
  home "#{node['splunk']['directory']}/splunkforwarder"
  system true
  shell '/bin/bash'
  notifies :create, "directory[#{node['splunk']['directory']}/splunkforwarder]", :immediate
end

#Disable THP
systemd_unit 'disable_thp.service' do
  content <<-EOU.gsub(/^\s+/, '')
  [Unit]
  Description=Disable Transparent Huge Pages (THP)

  [Service]
  Type=simple
  ExecStart=/bin/sh -c "echo 'never' > /sys/kernel/mm/transparent_hugepage/enabled && echo 'never' > /sys/kernel/mm/transparent_hugepage/defrag"

  [Install]
  WantedBy=multi-user.target
  EOU
  action [ :create, :enable, :start ]
end

#Configure ulimits for Splunk service account
limits_config "#{node['splunk']['user']}" do
  limits [
    { domain: "#{node['splunk']['user']}", type: 'hard', item: 'nproc', value: 16000 },
    { domain: "#{node['splunk']['user']}", type: 'hard', item: 'sigpending', value: 16000 },
    { domain: "#{node['splunk']['user']}", type: 'soft', item: 'sigpending', value: 16000 },
    { domain: "#{node['splunk']['user']}", type: 'hard', item: 'nofile', value: 64000 },
  ]
end
